File: src/arm_moveit_demo/scripts/04_cartesian.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import rospy, sys
from time import sleep
import moveit_commander
from copy import deepcopy
from moveit_commander import MoveGroupCommander, PlanningSceneInterface

if __name__ == "__main__":
    moveit_commander.roscpp_initialize(sys.argv)
    rospy.init_node('cartesian_plan_py')
    scene = PlanningSceneInterface()
    yahboomcar = MoveGroupCommander('arm_group')
    # 当运动规划失败后，允许重新规划
    yahboomcar.allow_replanning(True)
    yahboomcar.set_planning_time(50)
    yahboomcar.set_num_planning_attempts(20)
    yahboomcar.set_goal_position_tolerance(0.01)
    yahboomcar.set_goal_orientation_tolerance(0.01)
    yahboomcar.set_goal_tolerance(0.01)
    yahboomcar.set_max_velocity_scaling_factor(1.0)
    yahboomcar.set_max_acceleration_scaling_factor(1.0)
    rospy.loginfo("Set Init Pose")
    joints = [0, -1.57, -0.74, 0.71, 0]
    yahboomcar.set_joint_value_target(joints)
    yahboomcar.execute(yahboomcar.plan())
    sleep(0.5)
    # 获取当前位姿数据最为机械臂运动的起始位姿
    end_effector_link = yahboomcar.get_end_effector_link()
    start_pose = yahboomcar.get_current_pose(end_effector_link).pose
    # 初始化路点列表
    waypoints = []
    # 如果为True,将初始位姿加入路点列表
    waypoints.append(start_pose)
    for i in range(3):
        # 设置路点数据，并加入路点列表
        wpose = deepcopy(start_pose)
        wpose.position.z += 0.13
        waypoints.append(deepcopy(wpose))
        wpose.position.z -= 0.13
        waypoints.append(deepcopy(wpose))
    # 规划过程
    fraction = 0.0  # 路径规划覆盖率
    maxtries = 100  # 最大尝试规划次数
    attempts = 0    # 已经尝试规划次数
    rospy.loginfo("Path Planning in Cartesian Space")
    # 尝试规划一条笛卡尔空间下的路径，依次通过所有路点
    while fraction < 1.0 and attempts < maxtries:
        '''
        waypoints: 路点列表
        eef_step: 终端步进值，每隔0.1m计算一次逆解判断能否可达
        jump_threshold: 跳跃阈值，设置为0代表不允许跳跃
        plan: 路径, fraction: 路径规划覆盖率
        '''
        (plan, fraction) = yahboomcar.compute_cartesian_path(waypoints, 0.1, 0.0, True)
        attempts += 1
        if attempts % 10 == 0:
            rospy.loginfo("Still trying after " + str(attempts) + " attempts...")
    if fraction == 1.0:
        rospy.loginfo("Path computed successfully. Moving the yahboomcar.")
        yahboomcar.execute(plan)
        rospy.loginfo("Path execution complete.")
    else:
        rospy.loginfo("Path planning failed with only " + str(
            fraction) + " success after " + str(maxtries) + " attempts.")
    rospy.sleep(1)
    moveit_commander.roscpp_shutdown()
    moveit_commander.os._exit(0)

File: src/arm_moveit_demo/src/04_cartesian.cpp
#include <iostream>
#include "ros/ros.h"
#include <moveit/move_group_interface/move_group_interface.h>
#include <tf/LinearMath/Quaternion.h>
#include <moveit_visual_tools/moveit_visual_tools.h>

using namespace std;

int main(int argc, char **argv) {
    ros::init(argc, argv, "cartesian_plan_cpp");
    ros::NodeHandle n;
    ros::AsyncSpinner spinner(1);
    spinner.start();
    moveit::planning_interface::MoveGroupInterface yahboomcar("arm_group");
    string frame = yahboomcar.getPlanningFrame();
    moveit_visual_tools::MoveItVisualTools tool(frame);
    tool.deleteAllMarkers();
    yahboomcar.allowReplanning(true);
    // 规划的时间(单位：秒)
    yahboomcar.setPlanningTime(50);
    yahboomcar.setNumPlanningAttempts(10);
    // 设置允许目标角度误差
    yahboomcar.setGoalJointTolerance(0.01);
    yahboomcar.setGoalPositionTolerance(0.01);
    yahboomcar.setGoalOrientationTolerance(0.01);
    yahboomcar.setGoalTolerance(0.01);
    // 设置允许的最大速度和加速度
    yahboomcar.setMaxVelocityScalingFactor(1.0);
    yahboomcar.setMaxAccelerationScalingFactor(1.0);
    ROS_INFO("Set Init Pose.");
    //设置具体位置
    vector<double> pose{0, -0.69, -0.17, 0.86, 0};
    yahboomcar.setJointValueTarget(pose);
    sleep(0.5);
    moveit::planning_interface::MoveGroupInterface::Plan plan;
    yahboomcar.plan(plan);
    yahboomcar.execute(plan);
    // 获取当前位姿数据最为机械臂运动的起始位姿
    geometry_msgs::Pose start_pose = yahboomcar.getCurrentPose(yahboomcar.getEndEffectorLink()).pose;
    //初始化路径点向量
    std::vector<geometry_msgs::Pose> waypoints;
    //将初始位姿加入路点列表
    waypoints.push_back(start_pose);
    start_pose.position.x -= 0.04;
    waypoints.push_back(start_pose);
    start_pose.position.z -= 0.02;
    waypoints.push_back(start_pose);
    start_pose.position.x += 0.04;
    waypoints.push_back(start_pose);
    start_pose.position.z -= 0.02;
    waypoints.push_back(start_pose);
    start_pose.position.x += 0.03;
    waypoints.push_back(start_pose);
    // 笛卡尔空间下的路径规划
    moveit_msgs::RobotTrajectory trajectory;
    const double jump_threshold = 0.0;
    const double eef_step = 0.1;
    double fraction = 0.0;
    int maxtries = 1000;   //最大尝试规划次数
    int attempts = 0;     //已经尝试规划次数
    while (fraction < 1.0 && attempts < maxtries) {
        fraction = yahboomcar.computeCartesianPath(waypoints, eef_step, jump_threshold, trajectory);
        attempts++;
        if (attempts % 10 == 0) ROS_INFO("Still trying after %d attempts...", attempts);
    }
    if (fraction == 1) {
        ROS_INFO("Path computed successfully. Moving the arm.");
        // 生成机械臂的运动规划数据
        moveit::planning_interface::MoveGroupInterface::Plan plan;
        plan.trajectory_ = trajectory;
        // 显示轨迹
        tool.publishTrajectoryLine(plan.trajectory_, yahboomcar.getCurrentState()->getJointModelGroup("arm_group"));
        tool.trigger();
        // 执行运动
        yahboomcar.execute(plan);
        sleep(1);
    } else {
        ROS_INFO("Path planning failed with only %0.6f success after %d attempts.", fraction, maxtries);
    }
    return 0;
}

